portfolio/src/test/java/com/google/sps/CampaignServletTest.java

```diff
@@ -0,0 +1,162 @@
+
+import com.google.ads.googleads.lib.GoogleAdsClient;
+import com.google.ads.googleads.v3.services.GoogleAdsServiceClient;
+import com.google.auth.Credentials;
+import com.google.ads.googleads.v3.services.SearchGoogleAdsStreamRequest;
+import com.google.ads.googleads.v3.services.SearchGoogleAdsStreamResponse;
+
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+import static org.mockito.Matchers.anyInt;
+import static org.mockito.Matchers.anyString;
+import static org.powermock.api.mockito.PowerMockito.when;
+import static org.powermock.api.support.membermodification.MemberMatcher.method;
+
+
+import org.mockito.Spy;
+import org.mockito.InjectMocks;
+
+@PowerMockIgnore("jdk.internal.reflect.*")
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({DatastoreRetrieval.class, GetCampaignsServlet.class, SearchGoogleAdsStreamRequest.class,
+	SearchGoogleAdsStreamResponse.class, GoogleAdsServiceClient.class})
+public final class CampaignServletTest {
+
+  private GetCampaignsServlet servlet = PowerMockito.spy(new GetCampaignsServlet());
+  private TestCampaignsServlet testServlet;
+
+  @Before
+  public void setUp() {
+  	//servlet = PowerMockito.spy(new GetCampaignsServlet());
+  	testServlet = new TestCampaignsServlet();
+  	return;
+  }
+
+  @Test
+  public void mockTest() {
+  	HttpServletRequest request = mock(HttpServletRequest.class);       
+    HttpServletResponse response = mock(HttpServletResponse.class);  
+    when (request.getParameter("query")).thenReturn("SELECT campaign.id FROM campaign");
+
+    setDatastoreMocks();
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+
+    System.out.println(testServlet);
+    try {
+    	when(response.getWriter()).thenReturn(writer);
+    	testServlet.doPost(request, response);
+    } catch (Exception e) {
+    	System.out.println(e);
+    }
+    
+    verify(request, atLeast(1)).getParameter("query"); // make sure method is called
+    writer.flush();
+    System.out.println(stringWriter.toString());
+    //assertTrue(stringWriter.toString().contains("My expected string"));
+    Assert.assertEquals("Hello Ada", "Hello Ada");
+  }
+
+  private void setDatastoreMocks() {
+  	PowerMockito.mockStatic(DatastoreRetrieval.class);
+    PowerMockito.when(DatastoreRetrieval.getCredentialFromDatastore("DEVELOPER_TOKEN"))
+    	.thenReturn(System.getenv("DEVELOPER_TOKEN"));
+    PowerMockito.when(DatastoreRetrieval.getCredentialFromDatastore("CLIENT_ID"))
+    	.thenReturn(System.getenv("CLIENT_ID"));
+    PowerMockito.when(DatastoreRetrieval.getCredentialFromDatastore("CLIENT_SECRET"))
+    	.thenReturn(System.getenv("CLIENT_SECRET"));
+    PowerMockito.when(DatastoreRetrieval.getCredentialFromDatastore("refresh"))
+    	.thenReturn(System.getenv("REFRESH"));
+  }
+
+  class TestCampaignsServlet extends GetCampaignsServlet {
```

Comment:
This class seems a bit large and overflows the test. Maybe move it outside of the test class CampaignServletTest, or even better move it to a separate file. (1) This will make the test cleaner. (2) Also makes this mock reusable by other tests if any.